e2e/node/acp-v3.test.ts

```diff
@@ -0,0 +1,271 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import {
+  jest,
+  describe,
+  it,
+  expect,
+  beforeEach,
+  afterEach,
+} from "@jest/globals";
+
+import { Session } from "@inrupt/solid-client-authn-node";
+import { config } from "dotenv-flow";
+import {
+  getSolidDataset,
+  setThing,
+  saveSolidDatasetAt,
+  overwriteFile,
+  getFile,
+  getSourceUrl,
+  deleteFile,
+  createSolidDataset,
+  deleteSolidDataset,
+  UrlString,
+  acp_v3 as acp,
+  FetchError,
+} from "../../src/index";
+import { getTestingEnvironment, TestingEnvironment } from "./util/getTestingEnvironment";
+import { supportsAcp } from "./util/supportsAcp";
+import { getSession } from "./util/getSession";
+
+let env: TestingEnvironment;
+
+beforeAll(() => {
+  config({
+    path: __dirname,
+    // Disable warning messages in CI
+    silent: process.env.CI === "true",
+  });
+  env = getTestingEnvironment();
+});
+
+describe(`Authenticated end-to-end ACP V3 tests against environment ${env.environment}`, () => {
+  if (!supportsAcp(env.pod)) {
+    return;
+  }
+
+  let options: any;
+  let session: Session;
+  let sessionResource: string;
+
+  beforeEach(async () => {
+    session = await getSession(env);
+    sessionResource = `${env.pod}solid-client-tests/node/acp-v3-test-dataset-${session.info.sessionId}`;
+    options = { fetch: session.fetch };
+    await saveSolidDatasetAt(sessionResource, createSolidDataset(), options);
+  });
+
+  afterEach(async () => {
+    await deleteSolidDataset(sessionResource, options);
+    await session.logout();
+  });
+
+  async function initialisePolicyResource(
+    policyResourceUrl: UrlString,
+    session: Session
+  ) {
+    let publicRule = acp.createRule(policyResourceUrl + "#rule-public");
+    publicRule = acp.setPublic(publicRule);
+
+    let publicReadPolicy = acp.createPolicy(
+      policyResourceUrl + "#policy-publicRead"
+    );
+    // Note: we should think of a better name for "optional", as this isn't really optional.
+    //       At least one "optional" rule should apply, and since this is the only rule for this
+    //       policy, it will in practice be required.
+    publicReadPolicy = acp.addAnyOfRuleUrl(publicReadPolicy, publicRule);
+    publicReadPolicy = acp.setAllowModes(publicReadPolicy, {
+      read: true,
+      append: false,
+      write: false,
+    });
+
+    let selfRule = acp.createRule(policyResourceUrl + "#rule-self");
+    selfRule = acp.addAgent(selfRule, session.info.webId!);
```

Comment:
Checking that `session.info.webId` is not null could be added to the `beforeEach` block, and the non-null assertion could be replaced with a type assertion `as string`.

Reply:
Sure, but as for previous 2 comments, these are ACP V3 tests, so I will not refactor them as part of this PR.